### PR TITLE
Add a resource detector for Azure VMs

### LINKED ISSUE #5421: Add a resource detector for Azure VMs
### Problem Statement

Currently, there is no resource detector for Azure.

### Proposed Solution

A resource detector for Azure VMs should be provided which sets values according to semantic conventions for [host](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/host.md) and [cloud](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/cloud.md) resources.

### Alternatives
Setting relevant attributes manually without resource detectors.

### PR DESCRIPTION
**Pull request by pyohannes**
Fixes #5421 

This PR adds a resource detector for Azure VMs which sets values according to to semantic conventions for [host](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/host.md), [cloud](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/cloud.md), and [os](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/os.md) resources:

* `cloud.provider`
* `cloud.platform`
* `host.id`
* `host.name`
* `host.type`
* `os.type`
* `os.version`

### CODE REVIEW

File: detectors/azure/vm/README.md
Comment (maryliag) [inline detectors/azure/vm/README.md:23]: I see that you're also adding `cloud.region` and `cloud.resource_id`, but they're not listed here

**Comment by MrAlias** [comment]
@pyohannes we will want to add you as an owner for this.

**Comment by MrAlias** [comment]
> @pyohannes we will want to add you as an owner for this.

@pyohannes please go ahead and add this module [here](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/a520f281d284b4e1f8323f34e9b8d1fbcb9867eb/CODEOWNERS#L32-L33) with yourself and the go-approvers as the code owners.

File: detectors/azure/vm/README.md
Comment (pyohannes) [reply detectors/azure/vm/README.md:23]: Good catch, I added those.

**Comment by pyohannes** [comment]
> @pyohannes please go ahead and add this module [here](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/a520f281d284b4e1f8323f34e9b8d1fbcb9867eb/CODEOWNERS#L32-L33) with yourself and the go-approvers as the code owners.

@MrAlias Done, I added myself as a code owner.

File: CHANGELOG.md
Comment (dmathieu) [inline CHANGELOG.md:96]:
```suggestion
- Add the new `go.opentelemetry.io/contrib/detectors/azure/vm` package to provide a resource detector for Azure VMs. (#5422)
```

Comment (dmathieu) [reply CHANGELOG.md:96]: This will also need to be moved up to the unreleased section.

File: detectors/azure/vm/vm.go
Comment (dmathieu) [inline detectors/azure/vm/vm.go:124]: Since this method is mocked in the `Detect` test, could we test it anyway? (if we move the IP/port for the metadata endpoint into private variables, we can change them to an httptest endpoint in tests).

File: CHANGELOG.md
Comment (pyohannes) [reply CHANGELOG.md]: Done.

File: detectors/azure/vm/vm.go
Comment (pyohannes) [reply detectors/azure/vm/vm.go:124]: Yes, if so desired we can do that. We don't need to mock the method then, but just make the endpoint configurable.

Comment (pyohannes) [reply detectors/azure/vm/vm.go]: I removed the mockable client object, and instead made the endpoint URL configurable.

Please have another look @dmathieu and @MrAlias.

File: detectors/azure/vm/vm.go
Comment (dashpole) [inline detectors/azure/vm/vm.go:46]: In what cases would someone need to set this?  Or is this just for testing?

Do we still want to keep our parameters for version and format?

File: detectors/azure/vm/vm.go
Comment (dashpole) [inline detectors/azure/vm/vm.go:69]: Should we return a public struct rather than the resource.Detector interface?

File: detectors/azure/vm/vm.go
Comment (dashpole) [inline detectors/azure/vm/vm.go:75]: Detect should not return an error if it is not running on an azure VM.  It should only return errors if it is running on an azure VM, but fails for some reason to detect or process the response.  https://github.com/open-telemetry/opentelemetry-specification/blob/f3aeadbd152ce50c4d2cc9bc94902613bb49cffd/specification/resource/sdk.md?plain=1#L115

File: detectors/azure/vm/vm.go
Comment (dashpole) [inline detectors/azure/vm/vm.go:120]: nit: lowercase?
```suggestion
	pTransport := &http.Transport{Proxy: nil}

	client := http.Client{Transport: pTransport}
```

File: detectors/azure/vm/vm.go
Comment (dashpole) [inline detectors/azure/vm/vm.go:4]: `vm` doesn't seem like the best package name.  consider `azurevm`, or something that is likely to be relatively unique.

File: detectors/azure/vm/vm_test.go
Comment (pyohannes) [reply detectors/azure/vm/vm_test.go]: Good catch, this was a mistake. The intention is to use the status code from the test input.

File: detectors/azure/vm/vm.go
Comment (pyohannes) [reply detectors/azure/vm/vm.go]: Done. If the metadata server can't be reached then an empty resource is returned. If it returns invalid JSON, then an error is returned.

File: detectors/azure/vm/vm.go
Comment (pyohannes) [reply detectors/azure/vm/vm.go]: Changed it to `azurevm`.

I went with `vm` initially, as `detectors/azure/azurevm` seemed a bit redundant.

File: detectors/azure/vm/vm.go
Comment (pyohannes) [reply detectors/azure/vm/vm.go]: Whatever you  prefer.

I followed the example of [AWS](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/detectors/aws/ec2/ec2.go#L81) and [GCP](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/detectors/gcp/detector.go#L25) resource detectors, both return the `resource.Detector` interface.

File: detectors/azure/vm/vm.go
Comment (pyohannes) [reply detectors/azure/vm/vm.go]: This is for testing, I don't currently see another use case.

File: detectors/azure/vm/vm.go
Comment (dashpole) [reply detectors/azure/vm/vm.go]: I wrote the GCP detector, and should have used a struct.

File: detectors/azure/vm/vm.go
Comment (dashpole) [reply detectors/azure/vm/vm.go]: Lets not add a public function for `WithEndpoint` then.  You can still keep `Option` around for future extensibility. For testing, you can just set the endpoint field on the struct directly.

File: detectors/azure/azurevm/README.md
Comment (dashpole) [inline detectors/azure/azurevm/README.md:1]: Please convert this to a go doc: https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#documentation

File: detectors/azure/azurevm/vm.go
Comment (dashpole) [inline detectors/azure/azurevm/vm.go:78]: This is compliant with the spec (thanks!), but might hide real errors when someone is running on an azure vm.  Should we look for a more specific error before returning the empty resource?  E.g. if the metadata endpoint returns a 500, that probably means we are running on an azure VM, and we should return that error.

File: detectors/azure/vm/vm.go
Comment (pyohannes) [reply detectors/azure/vm/vm.go]: Done, I removed the options altogether.

File: detectors/azure/vm/vm.go
Comment (pyohannes) [reply detectors/azure/vm/vm.go]: Done, the struct is now returned.

File: detectors/azure/azurevm/vm.go
Comment (pyohannes) [reply detectors/azure/azurevm/vm.go:44]: I restructured the code a bit so that an empty resource is returned for HTTP client errors (4* status codes, which means we couldn't reach the server and can't be sure we're running in Azure). For all other cases, an error is returned (we can be quite confident we're running in Azure, and something went wrong).

File: detectors/azure/azurevm/README.md
Comment (pyohannes) [inline detectors/azure/azurevm/README.md:3]: This URL doesn't yet exist. Should the README only be added after the package is published?

File: detectors/azure/azurevm/vm.go
Comment (dashpole) [reply detectors/azure/azurevm/vm.go:45]: lgtm

File: detectors/azure/azurevm/vm.go
Comment (dashpole) [inline detectors/azure/azurevm/vm.go:44]: nit: not sure if it matters, but I would expect the context from Detect to be used with the http request.

**Review by dashpole — APPROVED** [review]

File: detectors/azure/azurevm/vm.go
Comment (MrAlias) [inline detectors/azure/azurevm/vm.go:15]:
```suggestion
	semconv "go.opentelemetry.io/otel/semconv/v1.25.0"
```

File: detectors/azure/azurevm/vm_test.go
Comment (MrAlias) [inline detectors/azure/azurevm/vm_test.go:17]:
```suggestion
	semconv "go.opentelemetry.io/otel/semconv/v1.25.0"
```

File: detectors/azure/azurevm/vm.go
Comment (MrAlias) [inline detectors/azure/azurevm/vm.go:20]: nit

```suggestion
const defaultAzureVMMetadataEndpoint = "http://169.254.169.254/metadata/instance/compute?api-version=2021-12-13&format=json"
```

File: detectors/azure/azurevm/vm.go
Comment (MrAlias) [inline detectors/azure/azurevm/vm.go:22]:
```suggestion
// ResourceDetector collects resource information of Azure VMs.
```

File: detectors/azure/azurevm/vm.go
Comment (MrAlias) [inline detectors/azure/azurevm/vm.go:37]:
```suggestion
// New returns a [ResourceDetector] that will detect Azure VM resources.
```

File: versions.yaml
Comment (MrAlias) [inline versions.yaml:23]: Do we want the first version release of this to be v0.53.0 or should this be its own modset and release it as v0.0.1?

Comment (pyohannes) [reply versions.yaml:23]: Either way is fine for me.

It seems that experimental groups are grouped by topic now (instrumentation, samplers, config, bridge, and processors), each with different versions.

Would you suggesting adding a new group, with version v0.0.1?

Comment (pyohannes) [reply versions.yaml]: @MrAlias Done, please have another look.

**Review by MrAlias — APPROVED** [review]

**Comment by MrAlias** [comment]
Blocked by https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5767

**Pull request merged**
